[1단계 - 장바구니 미션] 블루(이한희) 미션 제출합니다.

hanheel opened the pull request:
안녕하세요 쵸파! 블루라고 합니다.
이번 미션 리뷰 잘 부탁드립니다! :)

## 📦 장바구니 미션

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계

- 클라이언트 상태를 효과적으로 모델링하고 관리할 수 있다.
- Jest, React Testing Library(RTL)를 활용하여 주요 기능에 대한 테스트를 작성할 수 있다.

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [ ] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [ ] RTL 테스트 케이스를 모두 작성했나요?
- [ ] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?

  - 배포 링크 기입:  https://react-shopping-cart-zeta-lac.vercel.app/
 ⚠️ CORS 오류가 발생합니다!
 ⚠️ msw 모킹을 사용하였습니다

### 리뷰 요청 & 논의하고 싶은 내용

## ✅ 기능 요구사항
*  `/cart-items` API를 호출하여 장바구니 상품 데이터를 불러온다.
*  불러온 데이터를 기반으로 클라이언트 상태를 구성하고 관리한다.
    - 개별 상품의 선택 여부, 결제 금액, 배송비 등의 상태를 관리한다.
*  상품 선택에 따른 결제 금액, 배송비 등의 동적인 변경 사항을 처리한다.
    - 진입 시, 전체 선택 되어 있는 것이 디폴트이다.
    - 상품 선택/해제 시 결제 금액을 동적으로 변경한다.
    - 결제 금액이 10만원 이상일 경우 배송비는 무료이다.
*  장바구니 상품의 수량을 변경할 수 있다.
*  장바구니에 담긴 상품을 제거할 수 있다.


## 💬 의도 설명 
### 1️⃣ `useCheckedArray` 에  `useEffect` 가 포함된 이유
현재 장바구니 페이지에는 다음 두 가지의 훅이 존재합니다.
* `useCheckedArray` :  체크된 아이템을 관리하는 로직
* `useCartData` : 카트에 담긴 아이템을 관리하는 로직

처음 페이지가 렌더링될 때는 모든 아이템이 기본적으로 체크되어 있어야 하기 때문에, `useCheckedArray`는 `cartData`를 기반으로 초기 상태를 설정합니다. 따라서 `useCheckedArray`는 생성 시점에 `cartData`를 매개변수로 받아 내부 상태를 초기화합니다.

문제는 장바구니 아이템이 삭제될 경우, 기존에 체크되어 있던 아이템 중 해당 아이템도 함께 제거되어야 한다는 점입니다. 아이템 삭제 로직은 `useCartData` 내부에 있지만, 체크된 상태는 `useCheckedArray` 내부에 존재하기 때문에,  `useCartData` 가 `useCheckedArray`에 있는 체크된 아이템의 상태를 갖고 와야 합니다. 하지만 위와 같이 구현할 경우, 두 훅은 결국 쌍방향으로 의존하게 됩니다.

이를 해결하게 위해 `useCheckedArray` 내부에서 `useEffect`를 사용해 `cartData`의 변경을 감지하고 `checkedArray`를 갱신하는 방식으로 구현했습니다.

### 2️⃣ ToastContext 사용 이유 
현재 api를 호출시에 발생하는 오류는 `ToastContext` 를 통해 관리되고 있습니다.
* 토스트 메세지는 애플리케이션 전체에서 사용되는 UI 요소이다 - 어떤 컴포넌트에서든 쉽게 토스트 컴포넌트를 제어할 수 있어야 한다
* 토스트의 가시성, 메시지 내용, 성공/실패 상태를 한 곳에서 관리함으로써 **UI 상태의 일관성을 유지**하고, 중복 구현 없이 다양한 컴포넌트에서 재사용
위와 같은 이유로 `ToastContext`를 도입하였습니다!



## ❓ 질문
###  의도 설명 1번의 `useEffect` 사용 관련 질문
`useCartData`, `useCheckedArray` 두 훅의 쌍방향 의존성을 줄이기 위해 `useEffect` 를 사용했습니다. (의도설명 1번 내용과 동일합니다!)
* 위 의도를 갖고 `useEffect` 를 사용하는 것이 적절한 선택일까요?
* `useEffect` 를 사용하지 않는 방향으로 구조를 개선할 방법이 있을까요?

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 클라이언트 상태관리

- 원본상태/파생상태를 적절히 구분하여 선언하였나요?
- React state 를 불필요하게 선언한 부분은 없나요?
- 상태 관리 로직의 책임이 적절히 응집/분리되었나요? (ex. reducer, hook, Context)

### 2. MSW/Test

- 주요 기능을 적절히 정의하였나요?
- 주요 기능/예외에 대한 테스트가 충분히 이루어졌나요?

woowa-euijinkk committed:
chore: 의존성 제거

hanheel committed:
style: 디자인 시스템 정의

ExceptAnyone committed:
chore: emotion/react 세팅

ExceptAnyone committed:
chore: logo 이미지

ExceptAnyone committed:
chore: 기본 css 파일 삭제

ExceptAnyone committed:
feat: Header 컴포넌트 UI 구현

ExceptAnyone committed:
style: 글로벌 스타일 정의

ExceptAnyone committed:
feat: CartItem 컴포넌트 UI 구현

ExceptAnyone committed:
feat: 헤더 컴포넌트 삽입

hanheel committed:
style: cartItem 레이아웃 정리

ExceptAnyone committed:
feat: 화면 전체 Layout 구현

hanheel committed:
feat: cartItems get 호출

ExceptAnyone committed:
feat: Checkbox 컴포넌트 구현

ExceptAnyone committed:
feat: reponse Type 설정

ExceptAnyone committed:
feat: GET PATCH DELETE api 연결 함수 구현

ExceptAnyone committed:
feat: CartPage 페이지 구현

ExceptAnyone committed:
feat: 카트 아이템 업데이트 관련 props 전달

hanheel committed:
feat: 전체 체크 및 각 체크박스 체크 연결

ExceptAnyone committed:
feat: 삭제 로직 구현

ExceptAnyone committed:
feat: 총 결제 금액 추가

hanheel committed:
feat: msw 모킹 구현

ExceptAnyone committed:
chore: MSW 초기 설정

ExceptAnyone committed:
feat: 배송비 및 총 결제금액 계산

hanheel committed:
feat: 라우터 구현

hanheel committed:
feat: 주문완료 페이지 구현

ExceptAnyone committed:
chore: 라우터 설치

ExceptAnyone committed:
feat: 라우터  App 내 삽입

ExceptAnyone committed:
feat: useEasyNavigate 훅 구현

ExceptAnyone committed:
feat: 스타일 코드 분리

ExceptAnyone committed:
feat: 계산하는 유틸 함수들 구현

hanheel committed:
test: cartItem 테스트 작성

ExceptAnyone committed:
test: textEncoder 테스트 이슈 해결

hanheel committed:
feat: 장바구니에 상품이 없는 경우 메세지 렌더링

hanheel committed:
feat: 체크박스 default로 체크돼있게끔 변경

ExceptAnyone committed:
refactor: 로직들 커스텀 훅으로 분리

hanheel committed:
refactor: 상수 분리 및 중복 연산 분리

hanheel committed:
fix: 아이템 삭제시 버튼 disabled가 안되는 문제 해결

woowa-euijinkk committed:
docs: PR template

hanheel committed:
docs: PR 템플릿 삽입

hanheel committed:
feat: tryApiCall 유틸 함수 구현

ExceptAnyone committed:
feat: Toast 컴포넌트 구현

ExceptAnyone committed:
feat: useBoolean 커스텀 훅 구현

ExceptAnyone committed:
feat: tryApiCall 유틸 함수 구현

hanheel committed:
refactor: route를 상수로 관리

hanheel committed:
refactor: type import 설정

hanheel committed:
chore: 배포 404.html 해결

hanheel committed:
chore: 배포 msw 설정

hanheel committed:
fix: 배포 설정 파일명 오류 해결

ExceptAnyone committed:
chore: 콘솔로그 제거

ExceptAnyone committed:
refactor: 상수 코드들 하나의 객체로 정의 후 적용

hanheel committed:
refactor: 상수 구조화

hanheel committed:
refactor: 상수 구조화

bassyu commented on src/hooks/useCheckedArray.ts:
```diff
@@ -0,0 +1,47 @@
+import { useEffect, useState } from "react";
+import type { CartItemType } from "../types/response";
+
+const useCheckedArray = (cartData: CartItemType[]) => {
```
제네릭을 활용해서, 범용적인 훅으로 추출할 수도 있겠네요~!

hanheel replied:
반영했습니다! 👍🏻

✅ 커밋 해시 : [49a77a8](https://github.com/woowacourse/react-shopping-cart/pull/339/commits/49a77a81d714e8d4369f9c963c2e741a4f10a0ac)

bassyu commented on src/hooks/useCheckedArray.ts:
```diff
@@ -0,0 +1,47 @@
+const useCheckedArray = (cartData: CartItemType[]) => {
+  const [isCheckedArray, setIsCheckedArray] = useState<number[]>([]);
+
+  useEffect(() => {
+    initIsCheckedArray(cartData);
+  }, [cartData]);
```
> 위 의도를 갖고 useEffect 를 사용하는 것이 적절한 선택일까요?
useEffect 를 사용하지 않는 방향으로 구조를 개선할 방법이 있을까요?

상태 관리 로직이 워낙 범용적이어서, 처음 본 사람은 이 effect를 예측하기 힘들 것 같아요!
또는 추후에 cartData 상태를 임의로 변경하게 하게 된다면 이 effect로 인해서 복잡해질 것 같아요. (cartData가 서버 상태라 좋은 예시는 아닌 것 같네요..)

effect를 제거하는 방법으로는, `removeCartItem`이 성공한 이후에, `initIsCheckedArray`를 호출하면 되겠네요~!

hanheel replied:
말씀해 주신대로, 이 상황에서 useEffect를 사용하는 건 적절하지 못하다고 판단하여 다음과 같이 리팩토링했습니다!

useCheckedSet에서 useCartData를 참조하던 기존 흐름을 끊고,
반대로 `useCartData`에서 `syncIsCheckedSet` 함수를 직접 호출해 체크 상태를 명시적으로 동기화하는 방식을 선택했습니다.

`syncIsCheckedSet` 함수가 필요했던 이유는 다음과 같습니다
* 체크 상태(isCheckedSet)는 cartData와 별도로 유지됨
* 장바구니 아이템이 삭제되거나 갱신되면 cartData가 바뀌는데, 이때 기존 체크 상태에 존재하지 않는 ID가 남아 있을 수 있음
* 따라서 삭제 이후에는 존재하는 아이템 ID만 남기도록 동기화가 필요

✅ 커밋 : [1ebfdab](https://github.com/woowacourse/react-shopping-cart/pull/339/commits/1ebfdab563dcf9be4f4bec06c5c111194af5ce80)

hanheel replied:
질문이 있습니다..!

데이터가 삭제되거나 수정될 때마다 cartData와 isCheckedSet을 함께 갱신해줘야 해서
상태 동기화 로직이 반복적으로 발생하고, 구조가 점점 복잡해지는 느낌이 듭니다.

처음에는 cartData는 서버에서 내려온 데이터이고, isCheckedSet은 클라이언트에서만 사용하는 상태이기 때문에 분리해서 관리하는 것이 맞다고 판단했는데, 동기화를 매번 수동으로 맞춰줘야 하다 보니 이게 최선의 방식이 맞을까? 하는 의문이 듭니다.

상태 구조 자체를 개선해야 할 문제일까요? 아니면 동기화는 개발자가 꼼꼼히 챙겨야 하는 구조가 맞는 걸까요?

bassyu replied:
좋은 질문이네요! 말씀 주신 것처럼 `cartData`가 생성, 삭제될 때 `isCheckedSet`를 같이 바뀌어줘야 하는 번거로움이 있습니다.

두 상태가 다르기 때문에, CUD를 할 때 같이 동기화를 해줘야 합니다.
만약 비동기 서버 상태가 캡슐화 되어 있었다면, 초기 동기화를 위해 추가적인 `useEffect`가 들어가게 됩니다. 😢 

꼼꼼하게 챙겨야 하는 구조라고 생각해서, 
이 `업데이트 + 동기화` 동작들을 잘 묶어서 훅으로 만들고, 공유하는 방식을 추천합니다~!

bassyu commented on src/contexts/ToastContext.tsx:
> 토스트의 가시성, 메시지 내용, 성공/실패 상태를 한 곳에서 관리함으로써 UI 상태의 일관성을 유지하고, 중복 구현 없이 다양한 컴포넌트에서 재사용

좋습니다! 적절하게 사용해주신 것 같아요 👍 👍

bassyu commented on src/hooks/useBoolean.ts:
상태 관리 로직을 범용적으로 잘 분리 해주시네요!

bassyu commented on src/pages/cartPage/CartPage.tsx:
```diff
@@ -0,0 +1,106 @@
+      openToast,
+      '장바구니 데이터를 불러왔습니다.'
+    );
+    initIsCheckedArray(cartData);
```
API 반환 값의 `cartData`와 initIsCheckedArray가 알고 있는 `cartData`는 서로 다른 스코프의 상수인데, 의도된 것 맞을까요?
만약 아니라면, 네이밍을 다르게 해서 이런 상황을 예방하는 것을 추천합니다!

hanheel replied:
두 함수가 서로 다른 스코프를 참조 중이었군요...!! 

`getCart`로 받아온 `cartData`를 `initCartData`와 `initIsCheckedArray` 두 곳에서 공통으로 사용해야 하기 때문에,
API 호출 결과를 initialCartData로 저장해 한 번만 호출하고,
후속 초기화 함수에 각각 전달하는 방식으로 리팩토링했습니다!

```javascript
  useEffect(() => {
    const fetchCartData = async () => {
      const initialCartData = await tryApiCall(
        () => getCart(),
        openToast,
        "장바구니 데이터를 불러왔습니다."
      );
      if (!initialCartData) return;

      initCartData(initialCartData);
      initIsCheckedArray(initialCartData);
    };

    fetchCartData();
  }, []);
```

✅ 커밋 : [a9878fa](https://github.com/woowacourse/react-shopping-cart/pull/339/commits/a9878fa3357e1927d6077787a7d539f1ebcd4473)

bassyu commented on src/components/features/countButton/CountButton.tsx:
```diff
@@ -0,0 +1,35 @@
+      <button
+        css={S.countButtonButtonStyle}
+        onClick={() => increaseCartItem(cartId, quantity + 1)}
+      >
```
짧은 시간에 여러 번 클릭하면, API가 불필요하게 많이 호출되겠네요!
삭제 API 같은 케이스에서는 에러 메세지도 보일 것 같아요.

다양한 해결 법이 있을 것같은데, 어떻게 해결할 수 있을까요?

hanheel replied:
가장 쉽게 생각나는 방식은 플래그를 사용하는 방식인 것 같습니다!
```javascript
 const [isProcessing, setIsProcessing] = useState(false);

  const increaseCartItem = async (cartItemId: number, quantity: number) => {
    if (isProcessing) return;

    try {
      setIsProcessing(true);
//다른 로직
    } finally {
      setIsProcessing(false);
    }
  };

```

다만 위 방식으로 구현하게 되면 
1. 사용자에게는 무응답처럼 보일 수 있음
2. set함수는 비동기적으로 상태를 업데이트, isProcessing 상태에 반영되기 전에 연타할 경우 똑같이 서버에 불필요한 호출이 들어가게 됨
위와 같은 문제가 발생할 수 있다는 것을 깨달았습니다.

리뷰를 받고 공부해보니 외부 라이브러리를 사용하는 방식도 있다는 걸 새롭게 알게 됐습니다!
이번 미션에서는 외부 라이브러리 사용이 제한돼있기 때문에 우선 플래그를 사용하는 방식으로 구현하였습니다.
혹시 좀 더 효율적인 다른 방식도 있을까요? 🤔 

✅ 커밋 : [bda4383](https://github.com/woowacourse/react-shopping-cart/pull/339/commits/bda4383acfeb64c018b7dcc4b6f82b54a66aa3b7)

bassyu replied:
isProcessing일 때 버튼에 스피너를 넣는 등 표현해주면 어떨까요?

> isProcessing 상태에 반영되기 전에 연타할 경우

정말 빠르게 누르지 않는 이상, 이러기가 쉽지 않을 것 같긴 하네요~!

다른 방법으로는 API 함수 자체에 디바운싱을 거는 방법이 생각납니다!

bassyu commented on src/pages/cartPage/CartPage.tsx:
cartData가 로딩 중일 때, 그것을 표현해주는 UI가 있으면 좋을 것 같아요~!
마찬가지로 에러가 났을 때에도 단순히 토스트만 보여주는 것 이상으로, UI를 통해 더 친절하게 풀어주면 어떨까요?

hanheel replied:
1. 처음 장바구니 페이지에 접근할 때 - 스켈레톤 UI
<img width="150" alt="스크린샷 2025-06-03 오후 4 52 57" src="https://github.com/user-attachments/assets/29d6df18-89cb-4cb8-931c-ff37137df035" />



2. 장바구니 수량 조절 등 - opacity 조절
<img width="150" src="https://github.com/user-attachments/assets/b5d79320-6fa6-4879-80a6-48691d6bc4d0" />

3. 에러 발생시 

<img width="150" alt="스크린샷 2025-06-03 오후 5 02 12" src="https://github.com/user-attachments/assets/ce96042d-d99c-4f6e-96ba-adbc994ae51f" />



msw를 통해서 API 응답이 너무 빠르게 모킹되다 보니, 실제로 updating 상태를 눈으로 확인하기가 어려워, 
msw handler 함수들에 `delay`를 추가하였습니다!

✅ 커밋 : [b5e1694](https://github.com/woowacourse/react-shopping-cart/pull/339/commits/b5e1694d63283db81b8d3d1b5892321dfc853be7), [bec728c](https://github.com/woowacourse/react-shopping-cart/pull/339/commits/bec728cacb8bf68b3954e043fd0d1e1cd95cc066), [c130756](https://github.com/woowacourse/react-shopping-cart/pull/339/commits/c130756767551e41fdca14dbbc3f29fd9430e2a7)

- 최종 커밋 : [c130756](https://github.com/woowacourse/react-shopping-cart/pull/339/commits/c130756767551e41fdca14dbbc3f29fd9430e2a7)

bassyu commented on src/constants/errorMessages.ts:
상태에 따른 섬세한 에러 메세지 대응 좋네요 👍 👍

bassyu commented on src/pages/cartPage/CartPage.tsx:
로딩이나 에러 핸들링이 추가된다면, 이 핸들링이 잘 동작되는지 테스트로 추가해봐도 좋을 것 같아요!

hanheel replied:
반영했습니다! 


✅ 커밋 : [e774434](https://github.com/woowacourse/react-shopping-cart/pull/339/commits/e7744340c623bb7c48b76e4b739030185776f1c0)

bassyu commented on src/utils/calculate.ts:
```diff
@@ -0,0 +1,56 @@
+import type { CartItemType } from "../types/response";
+import { getCartItemById } from "./getCartItemById";
+
+export const getCartItemNamePrice = (
```
isCheckedArray가 배열이다보니, 여기 로직이 살짝 복잡해졌네요!
배열로 정의해주신 이유가 궁금합니다~!
 
checked 상태를 배열 대신 Set으로 관리해보면 어떨까요?

hanheel replied:
처음에는 총액 등 금액 관련 연산을 위해 checked 데이터를 사용해야 했고, 그 과정에서는 배열 메서드를 쓰는 것이 가장 간편했기 때문에 자연스럽게 선택했던 것 같습니다!

하지만 실제로 checked 상태를 관리할 때는 중복된 값을 신경 써야 하고, 포함 여부 판단이나 제거 등에서 로직이 복잡해지는 문제가 있었습니다. 말씀 주신대로 set으로 관리하니 체크관련 연산이 훨씬 간편해 진 것 같습니다! 

✅ 커밋 : [1ebfdab](https://github.com/woowacourse/react-shopping-cart/pull/339/commits/1ebfdab563dcf9be4f4bec06c5c111194af5ce80)

bassyu replied:
넵 좋습니다~! 배열은 순서를 가지는 데이터를 다룰 때 유용한 것 같아요!

bassyu commented on src/pages/cartPage/CartPage.tsx:
```diff
@@ -0,0 +1,106 @@
+            calculateTotalPrice(getCartItemNamePrice(isCheckedArray, cartData)),
+            calculateTotalProductCount(cartData, isCheckedArray)
+          )
+        }
```
최대한 가지고 있는 상태를 활용해 주신 점 좋네요! 👍 👍

bassyu commented on src/pages/cartPage/CartPage.tsx:
불필요한 상태 추가 없이, 원본 상태를 잘 활용해 주신 점 좋네요! 👍 👍

bassyu reviewed (changes requested):
안녕하세요 블루! 만나서 반갑습니다!
이번 미션 잘 부탁드려요~! 🙇 

코드를 일관성 있고 깔끔하게 작성해주셨네요! 덕분에 해석하기 좋았습니다 👍 👍 

기능도 큰 이슈 없이 잘 동작하네요~!
기능은 비동기 UI 핸들링 쪽을 보완 해주시면 더욱 좋을 것 같아요!

테스트는 조금 더 추가되면 좋을 것 같아요.
블루가 생각하는 중요 기능과 테스트 우선순위를 고민해서 정리 해주시고, 
우선순위가 높은 테스트부터 추가 해보면 좋겠습니다!

---

이번 미션도 수고 많으셨습니다!
코멘트 확인 후 다시 요청 주세요~! 💪 💪

hanheel committed:
refactor: useCheckedArray 범용적인 훅으로 재설계

1. 제네릭을 통해 데이터 형식을 입력받도록 설정
2. id 값이 필수로 포함되도록 설정

hanheel committed:
refactor: initCartData, initIsCheckedArray 같은 스코프의 값을 참조하도록 변경

hanheel committed:
feat: 이미지가 없을 경우 디폴트 이미지 로딩

hanheel committed:
refactor: cartPage 스타일 디테일 수정

hanheel committed:
feat: orderCompletePage 결제완료 버튼 추가 및 헤더 아이콘 렌더링

hanheel committed:
refactor: 애니메이션 파일 분리

hanheel committed:
feat: 로딩시 스켈레톤 설정

hanheel committed:
feat: 에러가 발생할 경우 에러 fallback 메세지를 UI에 출력

hanheel committed:
feat: useData 훅 구현 및 로딩 + 토스트 한 훅 내에서 관리

hanheel committed:
feat: 수량 update가 발생할 경우 opacity를 낮춤

hanheel committed:
fix: 삭제시 전체선택이 동기화되지 않는 문제 해결

1. useCheckedSet 내부에 syncIsCheckedSet 함수 생성
2. useCartData가 해당 함수를 통해 데이터를 동기화하도록 설정

hanheel committed:
test: 에러 / 로딩처리에 대한 테스트 추가

hanheel committed:
feat: 플래그를 통해 연속 API 호출 방지

hanheel committed:
fix: 버튼의 위치를 수정

hanheel committed:
fix: orderComplete 창에서 상품 종류의 개수 잘못 출력되는 문제 해결

bassyu reviewed (approved):
안녕하세요 블루!

리팩터링 잘 해주셨네요~!
다시 질문 주신 부분도 코멘트 드렸습니다!
제 의견이 정답은 아니니 블루 고민에 참고 해주세요!

1단계는 이만 머지하겠습니다!
2단계도 파이팅입니다~! 💪 💪
